File: crates/crypto-core/src/recovery.rs
//! Recovery phrase — wraps a copy of the user symmetric key.
//!
//! Wire-compatible with the web client (`clients/web/src/lib/crypto/recovery.ts`):
//!   * entropy: 32 random bytes
//!   * phrase: 64 lowercase hex chars shown as 8 space-separated groups of 8
//!   * recovery key: HKDF-Expand-SHA256(PRK = entropy, info = "nadsafe-recovery-v1", L = 64)
//!     — first 32 bytes = enc key, last 32 = mac key (no extract phase)
//!   * wrapped user key: standard type-2 EncString
//!
//! Both implementations are pinned to the same test vector below; change one
//! side only together with the other. Losing both master password and
//! recovery phrase = permanent vault loss (by design).

use hkdf::Hkdf;
use rand::RngCore;
use sha2::Sha256;
use zeroize::Zeroizing;

use crate::{
    enc_string::{EncString, SymmetricKey},
    error::CryptoError,
    keys::UserKey,
};

const RECOVERY_INFO: &[u8] = b"nadsafe-recovery-v1";

/// Generate fresh recovery entropy (32 random bytes).
pub fn generate_recovery_entropy() -> Zeroizing<[u8; 32]> {
    let mut entropy = Zeroizing::new([0u8; 32]);
    rand::thread_rng().fill_bytes(&mut *entropy);
    entropy
}

/// Encode entropy as the human-readable phrase: 8 groups of 8 hex chars.
pub fn entropy_to_phrase(entropy: &[u8; 32]) -> String {
    let hex: String = entropy.iter().map(|b| format!("{b:02x}")).collect();
    hex.as_bytes()
        .chunks(8)
        .map(|c| std::str::from_utf8(c).expect("hex is ASCII"))
        .collect::<Vec<_>>()
        .join(" ")
}

/// Parse a user-supplied phrase back to entropy. Whitespace-tolerant,
/// accepts upper- or lowercase hex.
pub fn phrase_to_entropy(phrase: &str) -> Result<Zeroizing<[u8; 32]>, CryptoError> {
    let hex: String = phrase.chars().filter(|c| !c.is_whitespace()).collect();
    if hex.len() != 64 || !hex.chars().all(|c| c.is_ascii_hexdigit()) {
        return Err(CryptoError::InvalidRecoveryPhrase);
    }
    let mut entropy = Zeroizing::new([0u8; 32]);
    for (i, byte) in entropy.iter_mut().enumerate() {
        *byte = u8::from_str_radix(&hex[i * 2..i * 2 + 2], 16)
            .map_err(|_| CryptoError::InvalidRecoveryPhrase)?;
    }
    Ok(entropy)
}

/// Derive the 64-byte recovery symmetric key from entropy.
/// HKDF-Expand only — the entropy is used directly as the PRK, matching the
/// web client's derivation (T(1) || T(2) for a 64-byte output).
fn recovery_key_from_entropy(entropy: &[u8; 32]) -> Result<Zeroizing<[u8; 64]>, CryptoError> {
    let hk = Hkdf::<Sha256>::from_prk(entropy).map_err(|e| CryptoError::Kdf(e.to_string()))?;
    let mut key = Zeroizing::new([0u8; 64]);
    hk.expand(RECOVERY_INFO, &mut *key)
        .map_err(|e| CryptoError::Kdf(e.to_string()))?;
    Ok(key)
}

/// Wrap the user's symmetric key with the recovery key.
/// Persist the returned EncString; show the phrase to the user once.
pub fn wrap_user_key_with_recovery(
    user_key: &UserKey,
    entropy: &[u8; 32],
) -> Result<EncString, CryptoError> {
    let recovery_key_bytes = recovery_key_from_entropy(entropy)?;
    let sym = SymmetricKey::from_bytes(&*recovery_key_bytes)?;
    EncString::encrypt(&user_key.raw_bytes(), &sym)
}

/// Unwrap the user's symmetric key using the recovery phrase.
/// Called during account recovery to restore vault access.
pub fn unwrap_user_key_with_recovery(
    enc_user_key: &EncString,
    phrase: &str,
) -> Result<UserKey, CryptoError> {
    let entropy = phrase_to_entropy(phrase)?;
    let recovery_key_bytes = recovery_key_from_entropy(&entropy)?;
    let sym = SymmetricKey::from_bytes(&*recovery_key_bytes)?;
    let raw = enc_user_key.decrypt(&sym)?;
    UserKey::from_bytes(&raw)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::kdf::{derive_master_key, Argon2idParams, Kdf};
    use crate::keys::MasterKey;

    fn test_user_key() -> UserKey {
        let kdf = Kdf::Argon2id(Argon2idParams {
            m_cost: 8192,
            t_cost: 1,
            p_cost: 1,
        });
        let mk_bytes = derive_master_key(b"password", "user@test.com", &kdf).unwrap();
        let mk = MasterKey(*mk_bytes);
        mk.generate_user_key().unwrap().0
    }

    /// Shared vector with the web client (recovery.test.ts) — both sides must
    /// derive these exact keys from entropy 0x00..0x1f or recovery blobs stop
    /// round-tripping between implementations.
    #[test]
    fn cross_implementation_vector() {
        let mut entropy = [0u8; 32];
        for (i, b) in entropy.iter_mut().enumerate() {
            *b = i as u8;
        }
        let key = recovery_key_from_entropy(&entropy).unwrap();
        assert_eq!(
            hex(&key[..32]),
            "8156d02754e1e4833d828086d0dddbecde161e86a6653fcbf8d3fc35e2255789",
        );
        assert_eq!(
            hex(&key[32..]),
            "e91ce477458041e028ebcba0ec20dddc70680ecf5f695295ce91f8a54c0fdce5",
        );
    }

    fn hex(bytes: &[u8]) -> String {
        bytes.iter().map(|b| format!("{b:02x}")).collect()
    }

    #[test]
    fn recovery_roundtrip() {
        let user_key = test_user_key();
        let entropy = generate_recovery_entropy();
        let phrase = entropy_to_phrase(&entropy);
        let wrapped = wrap_user_key_with_recovery(&user_key, &entropy).unwrap();

        let recovered = unwrap_user_key_with_recovery(&wrapped, &phrase).unwrap();
        assert_eq!(user_key.raw_bytes(), recovered.raw_bytes());
    }

    #[test]
    fn wrong_phrase_fails() {
        let user_key = test_user_key();
        let entropy = generate_recovery_entropy();
        let wrong_phrase = entropy_to_phrase(&generate_recovery_entropy());
        let wrapped = wrap_user_key_with_recovery(&user_key, &entropy).unwrap();

        assert!(unwrap_user_key_with_recovery(&wrapped, &wrong_phrase).is_err());
    }

    #[test]
    fn phrase_format_is_8_groups_of_8_hex() {
        let phrase = entropy_to_phrase(&generate_recovery_entropy());
        let groups: Vec<&str> = phrase.split(' ').collect();
        assert_eq!(groups.len(), 8);
        for g in groups {
            assert_eq!(g.len(), 8);
            assert!(g.chars().all(|c| c.is_ascii_hexdigit()));
        }
    }

    #[test]
    fn phrase_parse_roundtrip_and_validation() {
        let entropy = generate_recovery_entropy();
        let phrase = entropy_to_phrase(&entropy);
        assert_eq!(*phrase_to_entropy(&phrase).unwrap(), *entropy);
        // uppercase + extra whitespace tolerated
        assert_eq!(
            *phrase_to_entropy(&phrase.to_uppercase().replace(' ', "  ")).unwrap(),
            *entropy,
        );
        assert!(phrase_to_entropy("too short").is_err());
        assert!(phrase_to_entropy(&"zz".repeat(32)).is_err());
    }
}

File: clients/web/src/lib/crypto/recovery.ts
/**
 * Recovery phrase for vault recovery without master password.
 *
 * Flow:
 *   1. Generate 32 random bytes (recovery entropy)
 *   2. Derive recovery key via HKDF(entropy, "nadsafe-recovery-v1")
 *   3. Encrypt user key with recovery key (AES-256-CBC + HMAC)
 *   4. Store wrapped user key on server; user writes down the phrase
 *
 * Phrase format: 32 bytes displayed as 8 groups of 4 hex chars (256 bits).
 * Example: a1b2c3d4 e5f6a7b8 c9d0e1f2 ...
 */

import { encryptEncString, decryptEncString, serializeEncString, parseEncString } from "./enc-string";
import type { SymKey } from "./types";
import { symKeyFromBytes } from "./types";
import { toUtf8, randomBytes } from "./utils";

const RECOVERY_INFO = "nadsafe-recovery-v1";

/** Generate recovery phrase entropy (32 bytes). */
export function generateRecoveryEntropy(): Uint8Array {
  return randomBytes(32);
}

/**
 * Encode entropy as a human-readable phrase.
 * Format: 8 groups of 4 hex chars, space-separated.
 * 32 bytes → 64 hex chars → split into 8×8 hex → display as "a1b2c3d4 ..."
 */
export function entropyToPhrase(entropy: Uint8Array): string {
  const hex = Array.from(entropy)
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
  const groups: string[] = [];
  for (let i = 0; i < hex.length; i += 8) {
    groups.push(hex.slice(i, i + 8));
  }
  return groups.join(" ");
}

/** Parse recovery phrase back to raw entropy bytes. */
export function phraseToEntropy(phrase: string): Uint8Array {
  const hex = phrase.replace(/\s+/g, "");
  if (hex.length !== 64) throw new Error("Invalid recovery phrase length");
  if (!/^[0-9a-fA-F]{64}$/.test(hex)) throw new Error("Recovery phrase must be hexadecimal");
  const bytes = new Uint8Array(32);
  for (let i = 0; i < 32; i++) {
    bytes[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  }
  return bytes;
}

/** Derive a 64-byte recovery symmetric key from entropy via HKDF-Expand. */
async function deriveRecoveryKey(entropy: Uint8Array): Promise<{ encKey: Uint8Array; macKey: Uint8Array }> {
  // Copy to concrete ArrayBuffer to satisfy WebCrypto strict typing
  const entropyBuf = entropy.buffer.slice(entropy.byteOffset, entropy.byteOffset + entropy.byteLength) as ArrayBuffer;
  const prk = await crypto.subtle.importKey(
    "raw",
    entropyBuf,
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"],
  );

  const infoBytes = toUtf8(RECOVERY_INFO);

  // Expand two 32-byte blocks (HKDF-Expand, T(1) and T(2))
  const block1 = new Uint8Array(infoBytes.length + 1);
  block1.set(infoBytes);
  block1[infoBytes.length] = 0x01;
  const encKey = new Uint8Array(
    await crypto.subtle.sign("HMAC", prk, block1.buffer.slice(block1.byteOffset, block1.byteOffset + block1.byteLength) as ArrayBuffer),
  ).slice(0, 32);

  const block2 = new Uint8Array(32 + infoBytes.length + 1);
  block2.set(encKey);
  block2.set(infoBytes, 32);
  block2[32 + infoBytes.length] = 0x02;
  const macKey = new Uint8Array(
    await crypto.subtle.sign("HMAC", prk, block2.buffer.slice(block2.byteOffset, block2.byteOffset + block2.byteLength) as ArrayBuffer),
  ).slice(0, 32);

  return { encKey, macKey };
}

/**
 * Wrap user key with recovery key.
 * Returns an EncString — store on server for recovery.
 */
export async function wrapUserKeyWithRecovery(
  userKey: SymKey,
  recoveryEntropy: Uint8Array,
): Promise<string> {
  const { encKey, macKey } = await deriveRecoveryKey(recoveryEntropy);
  const userKeyBytes = new Uint8Array([...userKey.encKey, ...userKey.macKey]);
  const enc = await encryptEncString(userKeyBytes, { encKey, macKey });
  return serializeEncString(enc);
}

/**
 * Unwrap user key using recovery phrase.
 * Use this during recovery flow when master password is lost.
 */
export async function unwrapUserKeyWithRecovery(
  wrappedUserKey: string,
  recoveryPhrase: string,
): Promise<SymKey> {
  const entropy = phraseToEntropy(recoveryPhrase);
  const { encKey, macKey } = await deriveRecoveryKey(entropy);
  const parsed = parseEncString(wrappedUserKey);
  const raw = await decryptEncString(parsed, { encKey, macKey });
  return symKeyFromBytes(raw);
}

export type { SymKey };
